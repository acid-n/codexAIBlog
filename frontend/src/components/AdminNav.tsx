"use client";
import Link from "next/link";
import { useRouter } from "next/navigation";
import { PlusCircle, LogOut, User, Home, File } from "lucide-react";
import { useAuth } from "../contexts/AuthContext";
import Tooltip from "./Tooltip";

export default function AdminNav() {
  const { user, logout } = useAuth();
  const router = useRouter();
  if (!user) return null;
  const handleLogout = () => {
    logout();
    router.push("/");
  };
  return (
    <nav className="bg-gray-50 border-b py-1">
      <ul className="flex justify-center gap-4 text-gray-600 text-sm">
        <li>
          <Tooltip content={user.username}>
            <User aria-label="Профиль" className="w-5 h-5" />
          </Tooltip>
        </li>
        <li>
          <Tooltip content="Главная">
            <Link href="/" aria-label="Главная">
              <Home className="w-5 h-5" />
            </Link>
          </Tooltip>
        </li>
        <li>
          <Tooltip content="Создать пост">
            <Link href="/admin/create-post" aria-label="Создать пост">
              <PlusCircle className="w-5 h-5" />
            </Link>
          </Tooltip>
        </li>
        <li>
          <Tooltip content="Черновики">
            <Link href="/admin/drafts" aria-label="Черновики">
              <File className="w-5 h-5" />
            </Link>
          </Tooltip>
        </li>
        <li>
          <Tooltip content="Выход">
            <button onClick={handleLogout} aria-label="Выход">
              <LogOut className="w-5 h-5" />
            </button>
          </Tooltip>
        </li>
      </ul>
    </nav>
  );
}
