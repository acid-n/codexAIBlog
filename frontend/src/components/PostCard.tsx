"use client";
import Link from "next/link";
import { useAuth } from "../contexts/AuthContext";
import { Pencil } from "lucide-react";
import Tooltip from "./Tooltip";

export interface PostCardData {
  id: number;
  slug: string;
  title: string;
  description: string;
  author: number;
}

export default function PostCard({ post }: { post: PostCardData }) {
  const { user, isStaff } = useAuth();
  const canEdit = user && (user.id === post.author || isStaff);
  return (
    <article className="mx-auto max-w-prose text-center">
      <h2 className="mb-1 text-[28px] font-bold leading-snug">
        <Link href={`/posts/${post.slug}`} className="hover:underline">
          {post.title}
        </Link>
        {canEdit && (
          <Tooltip content="Редактировать">
            <Link href={`/posts/${post.slug}/edit`} aria-label="Редактировать">
              <Pencil className="inline-block w-4 h-4 ml-1 text-gray-500 hover:text-blue-600" />
            </Link>
          </Tooltip>
        )}
      </h2>
      <p className="mb-2 text-sm text-gray-500">
        20 августа 2024 · Автор · Категория
      </p>
      <p className="text-gray-700">{post.description}</p>
    </article>
  );
}
