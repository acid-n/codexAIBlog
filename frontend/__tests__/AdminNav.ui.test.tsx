import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AdminNav from "../src/components/AdminNav";
import PostCard from "../src/components/PostCard";

jest.mock("next/navigation", () => ({
  useRouter: () => ({ push: jest.fn() }),
}));

jest.mock("../src/contexts/AuthContext", () => ({
  useAuth: jest.fn(),
}));
const { useAuth } = require("../src/contexts/AuthContext");

describe("AdminNav", () => {
  test("guest has no panel", () => {
    useAuth.mockReturnValue({ user: null });
    const { container } = render(<AdminNav />);
    expect(container.firstChild).toBeNull();
  });

  test("renders panel with icons", () => {
    useAuth.mockReturnValue({
      user: { id: 1, username: "admin", email: "a@a", isStaff: false },
      logout: jest.fn(),
    });
    const { container } = render(<AdminNav />);
    expect(screen.getByRole("link", { name: /создать пост/i })).toHaveAttribute(
      "href",
      "/admin/create-post",
    );
    expect(container).toMatchSnapshot();
  });

  test("logout called", async () => {
    const logout = jest.fn();
    useAuth.mockReturnValue({
      user: { id: 1, username: "john", email: "", isStaff: false },
      logout,
    });
    const user = userEvent.setup();
    render(<AdminNav />);
    await user.click(screen.getByRole("button", { name: /выход/i }));
    expect(logout).toHaveBeenCalled();
  });
});

describe("PostCard edit icon", () => {
  test("shown only for author", () => {
    useAuth.mockReturnValue({
      user: { id: 1, username: "john", email: "", isStaff: false },
    });
    const post = {
      id: 1,
      slug: "test",
      title: "t",
      description: "d",
      author: 1,
    };
    const { unmount } = render(<PostCard post={post} />);
    expect(screen.getByLabelText(/редактировать/i)).toBeInTheDocument();

    unmount();
    const other = { ...post, author: 2 };
    render(<PostCard post={other} />);
    expect(screen.queryByLabelText(/редактировать/i)).toBeNull();
  });
});
