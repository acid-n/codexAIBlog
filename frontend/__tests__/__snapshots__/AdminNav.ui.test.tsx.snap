// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminNav renders panel with icons 1`] = `
<div>
  <nav
    class="bg-gray-50 border-b py-1"
  >
    <ul
      class="flex justify-center gap-4 text-gray-600 text-sm"
    >
      <li>
        <span
          class="relative inline-block group"
        >
          <svg
            aria-label="Профиль"
            class="lucide lucide-user w-5 h-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
            />
            <circle
              cx="12"
              cy="7"
              r="4"
            />
          </svg>
          <span
            class="pointer-events-none absolute left-1/2 top-full z-10 mt-1 hidden -translate-x-1/2 whitespace-nowrap rounded bg-black px-2 py-1 text-xs text-white group-hover:block"
          >
            admin
          </span>
        </span>
      </li>
      <li>
        <span
          class="relative inline-block group"
        >
          <a
            aria-label="Главная"
            href="/"
          >
            <svg
              class="lucide lucide-home w-5 h-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m3 9 9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
              />
              <polyline
                points="9 22 9 12 15 12 15 22"
              />
            </svg>
          </a>
          <span
            class="pointer-events-none absolute left-1/2 top-full z-10 mt-1 hidden -translate-x-1/2 whitespace-nowrap rounded bg-black px-2 py-1 text-xs text-white group-hover:block"
          >
            Главная
          </span>
        </span>
      </li>
      <li>
        <span
          class="relative inline-block group"
        >
          <a
            aria-label="Создать пост"
            href="/admin/create-post"
          >
            <svg
              class="lucide lucide-circle-plus w-5 h-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <path
                d="M8 12h8"
              />
              <path
                d="M12 8v8"
              />
            </svg>
          </a>
          <span
            class="pointer-events-none absolute left-1/2 top-full z-10 mt-1 hidden -translate-x-1/2 whitespace-nowrap rounded bg-black px-2 py-1 text-xs text-white group-hover:block"
          >
            Создать пост
          </span>
        </span>
      </li>
      <li>
        <span
          class="relative inline-block group"
        >
          <a
            aria-label="Черновики"
            href="/admin/drafts"
          >
            <svg
              class="lucide lucide-file w-5 h-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
              />
              <path
                d="M14 2v4a2 2 0 0 0 2 2h4"
              />
            </svg>
          </a>
          <span
            class="pointer-events-none absolute left-1/2 top-full z-10 mt-1 hidden -translate-x-1/2 whitespace-nowrap rounded bg-black px-2 py-1 text-xs text-white group-hover:block"
          >
            Черновики
          </span>
        </span>
      </li>
      <li>
        <span
          class="relative inline-block group"
        >
          <button
            aria-label="Выход"
          >
            <svg
              class="lucide lucide-log-out w-5 h-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
              />
              <polyline
                points="16 17 21 12 16 7"
              />
              <line
                x1="21"
                x2="9"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
          <span
            class="pointer-events-none absolute left-1/2 top-full z-10 mt-1 hidden -translate-x-1/2 whitespace-nowrap rounded bg-black px-2 py-1 text-xs text-white group-hover:block"
          >
            Выход
          </span>
        </span>
      </li>
    </ul>
  </nav>
</div>
`;
